Adoption of a new API for the DPIA compiler code

michel-steuwer opened the pull request:
This PR implements a Scala API based on "The Compiler Forest" and ported the DPIA compiler code to this API.

The corresponding paper is available here: https://homepages.inf.ed.ac.uk/gdp/publications/compiler-forest.pdf

The API described in this paper is implemented in the `util.compiler` package.

The C, OpenMP and OpenCL compilers implementations are changed to a more compositional design.
This PR is a preparation for allowing the adoption of the compiler forest API for the implementation of a host+GPU compiler.

Multiple additional improvements have been made, including:
 - a unified API in the `util.gen` object for initiating compilation to C, OpenMP, and OpenCL
 - introduction of the concept of `TranslationUnits` that encapsulate Functions / Kernels and their depending declarations
 - reorganisation of the former `"Program/KernelGeneration"` code
 - reorganising the package structure for DPIA primitives

Bastacyclop commented on src/main/scala/shine/DPIA/primitives/functional/AsScalar.scala:
```diff
@@ -12,19 +12,21 @@ import shine.DPIA._
+                           dt: ScalarType,
+                           access: AccessType,
+                           array: Phrase[ExpType]
+                         ) extends ExpPrimitive {
```
I am not a fan of all these indentation changes. The deleted code is more readble to me and takes less screen width.

michel-steuwer replied:
I have reverted these changes in the latest commit.

umazalakain commented on src/main/scala/exploration/runner/CExecutor.scala:
```diff
@@ -222,20 +223,20 @@ case class CExecutor(lowering: Strategy[Rise],
-    riseProgram.inputParams.foreach(elem => {
-      if (elem.`type`.dataType.toString.equals("int")) {
+    fun.inputParams.foreach{ case (elem, _) =>
+      if (elem.t.toString.equals("int")) {
```
Could this be made into `import shine.DPIA.Types.int; elem.t.equals(int)`?

Bastacyclop reviewed (changes requested):
Looks fine to me modulo my other comments and:
- Why `TranslationUnit`? According to a wikipedia search "A translation unit is the output of the C preprocessor – a source file after it has been preprocessed.". Maybe we could have a more intuitive name like `c/opencl.Declarations`?
- I am not in favor of merging all of the automatic spacing and style changes that your IDE triggered. In most cases I prefer the code style before the automatic changes and we started using scalastyle specifically because we wanted to avoid auto-formatting issues. I think that we should have the policy never to auto-format code, especially if the code has not actually been changed, and to configure IDEs accordingly.

michel-steuwer commented on src/main/scala/apps/harrisCornerDetection.scala:
```diff
@@ -301,11 +302,11 @@ object harrisCornerDetection {
+      gen.opencl.kernel.fromExpr("sobelY")(sobelY),
+      gen.opencl.kernel.fromExpr("mul")(mul),
+      gen.opencl.kernel.fromExpr("gaussian")(gaussian),
+      gen.opencl.kernel.fromExpr("coarsity")(coarsity)
```
I have adopted your suggestions and changed the API so that the kernel/function name (and the work group config) are now (optionally) given early. So the calls above now look like:
```
gen.opencl.kernel("sobelY").fromExpr(sobelY)
```
or
```
gen.opencl.kernel.fromExpr(sobelY)
```
when using a default name.

michel-steuwer commented on src/main/scala/shine/DPIA/Types/PhraseType.scala:
```diff
@@ -108,7 +109,7 @@ object PhraseType {
           case Identifier(name, _) =>
             if (`for`.name == name) {
-              Stop(NatAsIndex(ae.max, Natural(ae)).asInstanceOf[Phrase[T2]])
+              Stop(functional.NatAsIndex(ae.max, Natural(ae)).asInstanceOf[Phrase[T2]])
```
No. I removed this.

michel-steuwer commented on src/main/scala/shine/C/CodeGenerator.scala:
```diff
@@ -136,11 +100,11 @@ class CodeGenerator(val decls: CodeGenerator.Declarations,
       case Skip() => C.AST.Comment("skip")
 
-      case Seq(p1, p2) => C.AST.Stmts(cmd(p1, env), cmd(p2, env))
+      case Seq(p1, p2) => C.AST.Stmts(p1 |> cmd(env), p2 |> cmd(env))
```
No. I just chose this for consistency in this and related files.

michel-steuwer commented on src/main/scala/shine/DPIA/fromRise.scala:
```diff
@@ -10,6 +10,8 @@ import shine.DPIA.Phrases._
 import shine.DPIA.Types.DataType._
 import shine.DPIA.Types._
+import shine.DPIA.primitives.functional
+import shine.DPIA.primitives.functional.{AsScalar, AsVector, AsVectorAligned, Cast, DMatch, DepJoin, DepMapSeq, DepTile, Drop, ForeignFunctionCall, Fst, Gather, Generate, Idx, IndexAsNat, Iterate, IterateStream, Join, Let, MakeArray, MapFst, MapSeq, MapSeqUnroll, MapSnd, MapStream, MkDPair, NatAsIndex, Pad, PadClamp, PadEmpty, Pair, PrintType, ReduceSeq, ReduceSeqUnroll, Reorder, ScanSeq, Scatter ...
```
I have pushed a new commit that cleans up lots of these things and reverts most of the whitespace changes.

michel-steuwer commented on src/main/scala/shine/C/TranslationUnit.scala:
```diff
@@ -0,0 +1,25 @@
+case class TranslationUnit(decls: Seq[C.AST.Decl],
+                           functions: Seq[C.Function]) {
+  def compose(other: TranslationUnit): TranslationUnit =
+    TranslationUnit((decls ++ other.decls).distinct, functions ++ other.functions)
```
Duplicates are removed with the `.distinct` call for the declarations.
I don't think that we should / need to account for avoiding duplications of functions.

michel-steuwer commented on src/main/scala/shine/DPIA/primitives/functional/AsVector.scala:
```diff
@@ -12,15 +12,17 @@ import shine.DPIA._
+
+  array :: expT({
+    m * n
+  } `.` dt, access)
```
The latest commit reversed these changes.

michel-steuwer commented on src/test/scala/compiler/partialCompiler.scala:
```diff
@@ -0,0 +1,49 @@
+    type C_Compiler = Compiler[Phrase[_], C.TranslationUnit]
+    type OpenCL_Compiler = Compiler[Phrase[_], OpenCL.KernelTranslationUnit]
+  }
+}
```
That's right. This is only showing how the compiler forest API can be used.

michel-steuwer commented:
> * Why `TranslationUnit`? According to a wikipedia search "A translation unit is the output of the C preprocessor – a source file after it has been preprocessed.". Maybe we could have a more intuitive name like `c/opencl.Declarations`?

The `TranslationUnit` is supposed to capture one or multiple functions and declarations they depend on (e.g. declarations of structs, etc) - it is not (as the comment suggests) only a collection of declarations.
We called this before a `Program`, but I don't think this is a good term (we don't have a `main` function entry point or anything like this). A possible alternative term could be a `Module` - but I think I prefer Translation Unit. Opinions?

> * I am not in favor of merging all of the automatic spacing and style changes that your IDE triggered. In most cases I prefer the code style before the automatic changes and we started using scalastyle specifically because we wanted to avoid auto-formatting issues. I think that we should have the policy never to auto-format code, especially if the code has not actually been changed, and to configure IDEs accordingly.

I have reverted most of these changes. The refactoring of IDEA has triggered some auto formatting and auto import optimisation that resulted in sub-optimal results.

Bastacyclop commented on src/main/scala/shine/DPIA/primitives/functional/AsScalar.scala:
```diff
@@ -19,12 +19,12 @@ final case class AsScalar(
-  override val t: ExpType = expT({n * m}`.`dt, access)
+  array :: expT(n `.` vec(m, dt), access)
+  override val t: ExpType = expT({
+    n * m
```
nit: formatting

Bastacyclop commented on src/main/scala/shine/DPIA/primitives/functional/PrintType.scala:
```diff
@@ -1,19 +1,19 @@
+                           dt: DataType,
+                           access: AccessType,
+                           input: Phrase[ExpType]
+                          ) extends ExpPrimitive {
```
nit: formatting

Bastacyclop commented on src/main/scala/shine/DPIA/primitives/functional/Reorder.scala:
```diff
@@ -13,12 +13,12 @@ import shine.DPIA.{Phrases, _}
+                         idxF: Phrase[ExpType ->: ExpType],
+                         idxFinv: Phrase[ExpType ->: ExpType],
+                         input: Phrase[ExpType]
+                        ) extends ExpPrimitive {
```
nit: formatting

Bastacyclop commented on src/main/scala/shine/DPIA/primitives/functional/Take.scala:
```diff
@@ -18,14 +18,14 @@ final case class Take(
-    fun: VisitAndRebuild.Visitor
-  ): Phrase[ExpType] = {
+                                fun: VisitAndRebuild.Visitor
+                              ): Phrase[ExpType] = {
```
nit: formatting

Bastacyclop reviewed (commented):
> A possible alternative term could be a Module - but I think I prefer Translation Unit. Opinions?

Halide has a concept of "Module" "containing lowered function definitions and buffers." (https://halide-lang.org/docs/class_halide_1_1_module.html). I have a preference for `Module`.

> I have reverted most of these changes.

:+1:

michel-steuwer commented:
> > A possible alternative term could be a Module - but I think I prefer Translation Unit. Opinions?
> 
> Halide has a concept of "Module" "containing lowered function definitions and buffers." (https://halide-lang.org/docs/class_halide_1_1_module.html). I have a preference for `Module`.
> 
> > I have reverted most of these changes.
> 
> 👍

I renamed `TranslationUnit` to `Module` and fixed some last formatting issues. I kept some that I think are consistent.
Also, btw. I am preparing another PR that introduces macros for the DPIA patterns and will touch on their formatting anyway.

Thanks for your detailed comments @Bastacyclop.
I would appreciate if you could approve the PR or point out changes that you would like to see.

Bastacyclop reviewed (dismissed):
Looks good to me, can we merge https://github.com/rise-lang/shine/pull/82 first?

michel-steuwer commented:
Yes. I just merged #82 and fill fix the merge conflicts for them.
